File: device/cuda/elementwise/native/arithmetic.c
#include "arithmetic.h"
#include "elementwise.h"
#include "../internal/bridge/core_private.h"

#include <stdint.h>

static const char* cuda_binary_operation_name(int operation) {
    switch (operation) {
    case CUDABinaryFloat32Add:
        return "add";
    case CUDABinaryFloat32Sub:
        return "sub";
    case CUDABinaryFloat32Mul:
        return "mul";
    case CUDABinaryFloat32Div:
        return "div";
    case CUDABinaryFloat32Max:
        return "max";
    case CUDABinaryFloat32Min:
        return "min";
    default:
        return NULL;
    }
}

int cuda_dispatch_binary_elementwise(
    CUDADeviceRef contextRef,
    int operation,
    int elementDType,
    CUDABufferRef leftRef,
    CUDABufferRef rightRef,
    CUDABufferRef outRef,
    uint32_t count,
    uint64_t completionToken,
    CUDAStatus* status
) {
    cuda_elementwise_status_clear(status);

    if (count == 0) {
        return 0;
    }

    if (leftRef == NULL || rightRef == NULL || outRef == NULL) {
        cuda_elementwise_status_set(status, -2, "nil CUDA buffer");
        return -2;
    }

    const char* operationName = cuda_binary_operation_name(operation);

    if (operationName == NULL) {
        cuda_elementwise_status_set(status, -6, "unknown binary elementwise operation");
        return -6;
    }

    const char* dtypeSuffix = cuda_elementwise_element_dtype_suffix(elementDType);

    if (dtypeSuffix == NULL) {
        cuda_elementwise_status_set(status, -6, "unknown binary elementwise dtype");
        return -6;
    }

    char kernelName[128];
    int nameCode = cuda_elementwise_compose_kernel_name(
        kernelName,
        sizeof(kernelName),
        operationName,
        dtypeSuffix,
        status
    );

    if (nameCode != 0) {
        return nameCode;
    }

    const char* moduleSource = cuda_elementwise_module_source();

    if (moduleSource == NULL) {
        cuda_elementwise_status_set(status, -7, "CUDA elementwise module source not registered");
        return -7;
    }

    CUDAContext* context = NULL;
    CUDAStreamRef stream = NULL;
    int prepareCode = cuda_context_prepare(contextRef, status, &context, &stream);

    if (prepareCode != 0) {
        return prepareCode;
    }

    CUDAKernelRef kernel = cuda_get_kernel(context, moduleSource, kernelName, status);

    if (kernel == NULL) {
        return status != NULL && status->code != 0 ? status->code : -7;
    }

    void* leftPtr = cuda_buffer_device_ptr(leftRef);
    void* rightPtr = cuda_buffer_device_ptr(rightRef);
    void* outputPtr = cuda_buffer_device_ptr(outRef);
    void* args[] = {&leftPtr, &rightPtr, &outputPtr, &count};
    uint32_t launchCount = cuda_vector_launch_count(count, elementDType);
    int launchCode = cuda_launch_1d(context, kernel, stream, launchCount, args, sizeof(args), status);

    if (launchCode != 0) {
        return launchCode;
    }

    cuda_track_completion(context, stream, completionToken, NULL, status);
    return 0;
}

File: device/metal/elementwise/native/arithmetic.m
#include "arithmetic.h"
#include "elementwise.h"
#include "../internal/bridge/core_private.h"

#include <stdio.h>

static const char* metal_binary_operation_name(int operation) {
    switch (operation) {
    case MetalBinaryFloat32Add: return "add";
    case MetalBinaryFloat32Sub: return "sub";
    case MetalBinaryFloat32Mul: return "mul";
    case MetalBinaryFloat32Div: return "div";
    case MetalBinaryFloat32Max: return "max";
    case MetalBinaryFloat32Min: return "min";
    case MetalBinaryFloat32Eq: return "eq";
    case MetalBinaryFloat32Ne: return "ne";
    case MetalBinaryFloat32Lt: return "lt";
    case MetalBinaryFloat32Le: return "le";
    case MetalBinaryFloat32Gt: return "gt";
    case MetalBinaryFloat32Ge: return "ge";
    case MetalBinaryFloat32Pow: return "pow";
    case MetalBinaryFloat32Atan2: return "atan2";
    case MetalBinaryFloat32Mod: return "mod";
    default: return NULL;
    }
}

static int metal_elementwise_prepare(
    MetalContext* context,
    const char* kernelName,
    MetalStatus* status,
    id<MTLCommandQueue>* queue,
    id<MTLComputePipelineState>* pipeline
) {
    if (context == NULL || context->queue == NULL) {
        metal_elementwise_status_set(status, -1, "invalid Metal context");
        return -1;
    }

    *queue = (__bridge id<MTLCommandQueue>)context->queue;
    *pipeline = metal_get_pipeline(context, kernelName, status);

    if (*pipeline == nil) {
        return status != NULL && status->code != 0 ? status->code : -7;
    }

    return 0;
}

int metal_dispatch_binary_elementwise(
    MetalDeviceRef contextRef,
    int operation,
    int elementDType,
    MetalBufferRef leftRef,
    MetalBufferRef rightRef,
    MetalBufferRef outRef,
    uint32_t count,
    uint64_t completionToken,
    MetalStatus* status
) {
    @autoreleasepool {
        metal_elementwise_status_clear(status);

        if (count == 0) {
            return 0;
        }

        if (leftRef == NULL || rightRef == NULL || outRef == NULL) {
            metal_elementwise_status_set(status, -2, "nil Metal buffer");
            return -2;
        }

        char kernelName[128];
        int nameCode = metal_elementwise_compose_kernel_name(
            kernelName,
            sizeof(kernelName),
            metal_binary_operation_name(operation),
            metal_elementwise_element_dtype_suffix(elementDType),
            status
        );

        if (nameCode != 0) {
            return nameCode;
        }

        id<MTLCommandQueue> queue = nil;
        id<MTLComputePipelineState> pipeline = nil;
        int prepareCode = metal_elementwise_prepare(
            (MetalContext*)contextRef,
            kernelName,
            status,
            &queue,
            &pipeline
        );

        if (prepareCode != 0) {
            return prepareCode;
        }

        id<MTLCommandBuffer> commandBuffer;
        id<MTLComputeCommandEncoder> encoder = metal_get_encoder((MetalContext*)contextRef, &commandBuffer);

        [encoder setComputePipelineState:pipeline];
        [encoder setBuffer:(__bridge id<MTLBuffer>)leftRef offset:0 atIndex:0];
        [encoder setBuffer:(__bridge id<MTLBuffer>)rightRef offset:0 atIndex:1];
        [encoder setBuffer:(__bridge id<MTLBuffer>)outRef offset:0 atIndex:2];
        [encoder setBytes:&count length:sizeof(count) atIndex:3];

        NSUInteger threadWidth = [pipeline threadExecutionWidth];

        if (threadWidth == 0) {
            threadWidth = 1;
        }

        NSUInteger vectorCount = (NSUInteger)((count + 3) / 4);
        [encoder
            dispatchThreads:MTLSizeMake(vectorCount, 1, 1)
            threadsPerThreadgroup:MTLSizeMake(threadWidth, 1, 1)
        ];
        metal_track_command_completion((MetalContext*)contextRef, commandBuffer, completionToken, NULL);
        metal_end_encoder((MetalContext*)contextRef, encoder, commandBuffer);

        return 0;
    }
}
